File: src/io/github/hulang1024/chinesechess/scene/chessplay/rule/chess/ChessN.java
package io.github.hulang1024.chinesechess.scene.chessplay.rule.chess;

import io.github.hulang1024.chinesechess.scene.chessplay.rule.ChessPosition;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.RoundGame;

/**
 * 马
 * @author Hu Lang
 */
public class ChessN extends AbstractChess {
    @Override
    public boolean canGoTo(ChessPosition destPos, RoundGame game) {
        // 马走“日”，蹩马腿
        int rowOffset = destPos.row - pos.row;
        int colOffset = destPos.col - pos.col;

        if (Math.abs(rowOffset) == 2 && Math.abs(colOffset) == 1) {
            return game.chessboard.isEmpty(pos.row + (rowOffset > 0 ? +1 : -1), pos.col);
        } else if (Math.abs(rowOffset) == 1 && Math.abs(colOffset) == 2) {
            return game.chessboard.isEmpty(pos.row, pos.col + (colOffset > 0 ? +1 : -1));
        } else {
            return false;
        }
    }
}

File: client/src/rulesets/chinesechess/rule/ChessN.ts
import ChessPos from "./ChessPos";
import AbstractChess from "./AbstractChess";
import { sign } from "./move_rules";
import ChessboardState from "./ChessboardState";

/**
 * 马
 */
export default class ChessN extends AbstractChess {
  canGoTo(destPos: ChessPos, chessboardState: ChessboardState): boolean {
    // 马走“日”，蹩马腿
    const rowOffset = destPos.row - this.pos.row;
    const colOffset = destPos.col - this.pos.col;

    if (Math.abs(rowOffset) == 2 && Math.abs(colOffset) == 1) {
      return chessboardState.isEmpty(this.pos.row + sign(rowOffset), this.pos.col);
    }
    if (Math.abs(rowOffset) == 1 && Math.abs(colOffset) == 2) {
      return chessboardState.isEmpty(this.pos.row, this.pos.col + sign(colOffset));
    }
    return false;
  }

  clone(): ChessN {
    const clone = new ChessN(this.pos, this.host);
    clone.front = this.front;
    return clone;
  }
}
